### PR TITLE
Visualizable Mixin

### PR DESCRIPTION
**Pull request by ihowell**
Adds the visualizable mixin and documentation around it.

### CODE REVIEW

File: pyroclast/common/visualizable.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+import tensorflow as tf
+
+
+class VisualizableMixin(abc.ABC):
+    """Interface Model interface.
+
+    Provides functionality for various calculations and
+    visualizations. Should be used as a mixin in conjunction with
+    tf.Module.
+
```
Comment (DrKwint) [inline pyroclast/common/visualizable.py]: why the space here?

File: pyroclast/common/visualizable.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    """
+
+    @abc.abstractmethod
+    def classify(self, x):
+        """Calculates output logits given a set of inputs.
```
Comment (DrKwint) [inline pyroclast/common/visualizable.py]: what is a "set of inputs"?

File: pyroclast/common/visualizable.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+           x (tf.Tensor): Input of shape [batch_size, ...data_shape]
+
+        Returns:
+            logits (tf.Tensor): shape [batch_size, num_classes]
+        """
+        pass
+
+    @abc.abstractmethod
+    def conv_stack_submodel(self):
+        """Returns the submodel that contains the convolution stack
```
Comment (DrKwint) [inline pyroclast/common/visualizable.py]: Does it return an object which contains the conv stack, or does it return the conv stack?

File: pyroclast/common/visualizable.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+        This method is used to retrieve the conv stack of a network in
+        order to find activation maps.
+
+        Returns:
+           submodel (tf.Module): A submodel that begins at the first layer of the network and includes the conv stack.
+
```
Comment (DrKwint) [inline pyroclast/common/visualizable.py]: again, why space here?

File: pyroclast/common/visualizable.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    def cam_map(self, x):
+        """Calculates the class activation mapping
+        https://http://cnnlocalization.csail.mit.edu/
+
+        Args: x (np.array): shape [batch_size, ...data_shape]
+
+        Returns:
+           activation_map (np.array): shape [batch_size, ...data_shape]
+
+        TODO: This
```
Comment (DrKwint) [inline pyroclast/common/visualizable.py]: TODO

File: pyroclast/common/visualizable.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        Returns:
+           activation_map (np.array): shape [batch_size, ...data_shape]
+
+        """
+        result = self.conv_stack_submodel()(x)
+        return tf.image.resize(result, [x.shape[1], x.shape[2]])
+
+    def cam_map(self, x):
+        """Calculates the class activation mapping
+        https://http://cnnlocalization.csail.mit.edu/
```
Comment (DrKwint) [inline pyroclast/common/visualizable.py]: What is this link to?

File: pyroclast/common/visualizable_test.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+            y = self.model(x)
+            assert y.shape == (self.args['batch_size'], 10)
+
+    def test_activation_map(self):
+        for batch in self.ds['train']:
+            x = tf.cast(batch['image'], tf.float32)
+            activation_maps = self.model.activation_map(x)
+            assert activation_maps.shape[0:3] == x.shape[0:3]
+
+    def test_cam_map(self):
```
Comment (DrKwint) [inline pyroclast/common/visualizable_test.py]: Should this test be empty?

File: pyroclast/common/visualizable.py (hunk 2 above)
Comment (ihowell) [reply pyroclast/common/visualizable.py]: True. This should probably just be a tensor.

File: pyroclast/common/visualizable.py (hunk 3 above)
Comment (ihowell) [reply pyroclast/common/visualizable.py]: I think that it should contain the final layer in the conv stack. What do you think?

File: pyroclast/common/visualizable.py (hunk 6 above)
Comment (ihowell) [reply pyroclast/common/visualizable.py]: Lol, besides the double http:// it links to the cam paper website.

File: pyroclast/common/visualizable.py (hunk 5 above)
Comment (ihowell) [reply pyroclast/common/visualizable.py]: Yes, we determined that we would do this later, when it is actually needed.

File: pyroclast/common/visualizable_test.py (hunk 7 above)
Comment (ihowell) [reply pyroclast/common/visualizable_test.py]: I just wanted to leave it in as a reminder to do the test once cam_map is implemented.

**Pull request merged**